dask/array/core.py

```diff
@@ -2596,7 +2558,70 @@ def reshape(array, shape):
 
     if np.prod(shape) != array.size:
```

Comment:
Since this should be in before the next release, and I *think* this is good to go, I'd prefer to merge the new reshape, and then fix it in #1478. A few more calls to `np.prod` are introduced here that should also be changed.